homeassistant/helpers/event.py

```diff
@@ -349,41 +353,323 @@ def remove_listener() -> None:
+    Callable to unregister the listener.
 
-    # Local variable to keep track of if the action has already been triggered
-    already_triggered = False
+    """
 
     @callback
-    def template_condition_listener(entity_id: str, from_s: State, to_s: State) -> None:
+    def state_changed_listener(
+        event: Event,
+        template: Template,
+        last_result: Optional[str],
+        result: Union[str, TemplateError],
+    ) -> None:
         """Check if condition is correct and run action."""
-        nonlocal already_triggered
-        template_result = condition.async_template(hass, template, variables)
-
-        # Check to see if template returns true
-        if template_result and not already_triggered:
-            already_triggered = True
-            hass.async_run_job(action, entity_id, from_s, to_s)
-        elif not template_result:
-            already_triggered = False
-
-    return async_track_state_change(
-        hass, template.extract_entities(variables), template_condition_listener
+        if isinstance(result, TemplateError):
+            _LOGGER.exception(result)
+            return
+
+        if _boolean_coerce(last_result) or not _boolean_coerce(result):
+            return
+
+        hass.async_run_job(
+            action,
+            event.data.get("entity_id"),
+            event.data.get("old_state"),
+            event.data.get("new_state"),
+        )
+
+    info = async_track_template_result(
+        hass, template, state_changed_listener, variables
     )
 
+    return info.async_remove
+
 
 track_template = threaded_listener_factory(async_track_template)
 
 
+_UNCHANGED = object()
+
+
+class TrackTemplateResultInfo:
+    """Handle removal / refresh of tracker."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        template: Template,
+        action: Callable,
+        variables: Optional[TemplateVarsType],
+    ):
+        """Handle removal / refresh of tracker init."""
+        self.hass = hass
+        self._template = template
+        self._action = action
+        self._variables = variables
+        self._last_result: Optional[str] = None
+        self._last_exception = False
+        self._all_listener: Optional[Callable] = None
+        self._domains_listener: Optional[Callable] = None
+        self._entities_listener: Optional[Callable] = None
+        self._info = template.async_render_to_info(variables)
+        if self._info.exception:
+            self._last_exception = True
+            _LOGGER.exception(self._info.exception)
+        self._create_listeners()
+        self._last_info = self._info
+
+    @property
+    def _needs_all_listener(self) -> bool:
+        # Tracking all states
+        if self._info.all_states:
+            return True
+
+        # Previous call had an exception
+        # so we do not know which states
+        # to track
+        if self._info.exception:
+            return True
+
+        # There are no entities in the template
```

Comment:
For a future PR, we should collect use cases and see when this is used. People doing this seems like a big hack (check time or something) and it doesn't scale with many state change listeners.

Reply:
Maybe we should log a warning once when the template is setup?

Reply:
When I start working on the template integrations, I'll have a better idea on what the impact is here.

Noted for the future